File: 백준/Gold/2493. 탑/탑.cc
#include <iostream>
#include <stack>
#include <utility>

using namespace std;

int n, input;
stack<pair<int, int>> st;

int main(void) {
	cin.tie(0)->sync_with_stdio(false);

	cin >> n;

	for (int i = 0; i < n; i++) {
		cin >> input;
		
		while (!st.empty()) {
			if (input < st.top().second) {
				cout << st.top().first + 1 << " ";
				st.push(make_pair(i, input));
				break;
			}
			else {
				st.pop();
			}
		}

		if (st.empty()) {
			cout << "0 ";
			st.push(make_pair(i, input));
		}
	}

	return 0;
}

File: 백준/Gold/2493. 탑/탑.py
n = int(input())

stack = []
height = list(map(int, input().split()))

top_list = []

for i in range(len(height)):
    top_list.append({ 'height' : height[i], 'id' : i + 1})

for top in top_list:
    while stack:
        if stack[-1]['height'] >= top['height']:
            print(stack[-1]['id'], end=' ')
            stack.append(top)
            break
        else:
            stack.pop()

    if not stack:        
        stack.append(top)
        print(0, end=' ')